### PR TITLE
Protect creation/destruction of boost's named_mutex [16273]

### PR DESCRIPTION
**Pull request by juanlofer-eprosima**
Thread sanitizer reports a data race when attempting to create a boost's named_mutex for the same port from different threads (in the same process). This situation occurs, for example, when a two participants in the same process discover another one from a different process at the same time. Both attempt to create and lock a named_mutex corresponding to the same port (remote participant's) through boost library, which seems not to be thread safe according to thread sanitizer. Likewise, opening and creating a named_mutex for the same port from different threads results in a data race being reported. Same goes for destroying and opening/creating a named_mutex.

This PR attempts to solve this data race by protecting the creation/destruction of a named_mutex with a static mutex in `SharedMemSegment`. Note that this solution is sub-optimal, since now there would be blocking when trying to create any two named_mutexes concurrently, even if they do not correspond to the same port.

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added. 
- [ ] Any new/modified methods have been properly documented using Doxygen. 
- [ ] Fast DDS test suite has been run locally. 
- [x] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- [x] Documentation builds and tests pass locally. 
- [ ] New feature has been added to the `versions.md` file (if applicable).
- [ ] New feature has been documented/Current behavior is correctly described in the documentation. 

## Reviewer Checklist
- [ ] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Review by jsantiago-eProsima — CHANGES_REQUESTED** [review]
Some minor changes. Also, please uncrustify.

**Comment by jsantiago-eProsima** [comment]
@richiprosima please test this

File: test/unittest/transport/SharedMemTests.cpp [hunk 1]
```diff
@@ -2161,6 +2161,108 @@ TEST_F(SHMTransportTests, dump_file)
...
+    Semaphore sem_second_lock_done;
+    Semaphore sem_end_thread_locker;
+    std::atomic<int> lock_count(0);
+    std::thread thread_locker([&]
+            {
+                // lock has to be done in another thread because
+                // boost::inteprocess_named_mutex and  interprocess_mutex are recursive in Win32
+                auto port_mutex = port_mocker.get_port_mutex(domain_name, 0);
+                bool locked;
+                if ((locked = port_mutex->try_lock()))
```
Comment (MiguelCompany) [inline test/unittest/transport/SharedMemTests.cpp:2215]:
```suggestion
                bool locked = port_mutex->try_lock();
                if (locked)
```

File: test/unittest/transport/SharedMemTests.cpp [hunk 2]
```diff
@@ -2161,6 +2161,108 @@ TEST_F(SHMTransportTests, dump_file)
...
+                }
+
+                sem_second_lock_done.post();
+                sem_end_thread_locker.wait();
+            }
+            );
+
+    auto port_mutex = port_mocker.get_port_mutex(domain_name, 0);
+    bool locked;
+    if ((locked = port_mutex->try_lock()))
```
Comment (MiguelCompany) [inline test/unittest/transport/SharedMemTests.cpp:2240]:
```suggestion
    bool locked = port_mutex->try_lock();
    if (locked)
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Fix warnings on windows with my suggestions

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Review by jsantiago-eProsima — APPROVED** [review]

**Comment by EduPonz** [comment]
@richiprosima please test this

**Comment by MiguelCompany** [comment]
CI for 2.9.0-rc:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/fastdds_linux/3187)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/fastdds_mac/2944)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/fastdds_windows/3349/)
* Discovery server [Build Status](http://jenkins.eprosima.com:8080/job/discovery_server_linux/3131/)

**Pull request merged**